File: Subsets/Subsets.cpp
class Solution {
public:
    vector<vector<int> > subsets(vector<int> &S) {
        sort(S.begin(), S.end());
        vector<vector<int> > res;
        for (int i = 0; i <= S.size(); ++i) {
            vector<vector<int> > tmp;
            vector<int> list;
            subsetsOfKHelper(tmp, list, 0, S, i);
            for (vector<int> vec : tmp) {
                res.push_back(vec);
            }
        }
        return res;
    }
    void subsetsOfKHelper(vector<vector<int> > &res, vector<int> &list, int start, vector<int> &S, int k) {
        if (list.size() == k) {
            res.push_back(vector<int>(list));
            return;
        }
        for (int i = start; i < S.size(); ++i) {
            list.push_back(S[i]);
            subsetsOfKHelper(res, list, i + 1, S, k);
            list.pop_back();
        }
    }
};

File: Subsets/Subsets.java
public class Solution {
    public List<List<Integer>> subsets(int[] S) {
        Arrays.sort(S);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        for (int i = 0; i <= S.length; ++i) {
            List<List<Integer>> tmp = new ArrayList<List<Integer>>();
            subsetsOfKHelper(tmp, new ArrayList<Integer>(), 0, S, i);
            res.addAll(tmp);
        }
        return res;
    }
    private void subsetsOfKHelper(List<List<Integer>> res, List<Integer> list, int start, int[] S, int k) {
        if (list.size() == k) {
            res.add(new ArrayList(list));
            return;
        }
        for (int i = start; i < S.length; ++i) {
            list.add(S[i]);
            subsetsOfKHelper(res, list, i + 1, S, k);
            list.remove(list.size() - 1);
        }
    }
}